cli/cmd/inject.go

```diff
@@ -209,105 +214,148 @@ func InjectYAML(in io.Reader, out io.Writer, version string) error {
-		//     between containers in the same pod.
-		//  3. Using a Service IP instead of localhost would mean intra-pod traffic would be load-balanced
-		//     across all controller pod replicas. This is undesirable as we would want all traffic between
-		//	   containers to be self contained.
-		//  4. We skip recording telemetry for intra-pod traffic within the control plane.
-		switch meta.Kind {
-		case "Deployment":
-			var deployment v1beta1.Deployment
-			err = yaml.Unmarshal(bytes, &deployment)
-			if err != nil {
-				return err
-			}
-			if deployment.Name == ControlPlanePodName && deployment.Namespace == controlPlaneNamespace {
-				DNSNameOverride = LocalhostDNSNameOverride
-			}
-			obj = &deployment
-			k8sLabels[k8s.ProxyDeploymentLabel] = deployment.Name
-			podTemplateSpec = &deployment.Spec.Template
-		case "ReplicationController":
-			var rc v1.ReplicationController
-			err = yaml.Unmarshal(bytes, &rc)
-			if err != nil {
-				return err
-			}
-			obj = &rc
-			k8sLabels[k8s.ProxyReplicationControllerLabel] = rc.Name
-			podTemplateSpec = rc.Spec.Template
-		case "ReplicaSet":
-			var rs v1beta1.ReplicaSet
-			err = yaml.Unmarshal(bytes, &rs)
-			if err != nil {
-				return err
-			}
-			obj = &rs
-			k8sLabels[k8s.ProxyReplicaSetLabel] = rs.Name
-			podTemplateSpec = &rs.Spec.Template
-		case "Job":
-			var job batchV1.Job
-			err = yaml.Unmarshal(bytes, &job)
-			if err != nil {
-				return err
-			}
-			obj = &job
-			k8sLabels[k8s.ProxyJobLabel] = job.Name
-			podTemplateSpec = &job.Spec.Template
-		case "DaemonSet":
-			var ds v1beta1.DaemonSet
-			err = yaml.Unmarshal(bytes, &ds)
-			if err != nil {
-				return err
-			}
-			obj = &ds
-			k8sLabels[k8s.ProxyDaemonSetLabel] = ds.Name
-			podTemplateSpec = &ds.Spec.Template
+		out.Write(result)
+		out.Write([]byte("---\n"))
+	}
+
+	return nil
+}
+
+func injectList(b []byte, version string) ([]byte, error) {
+	var sourceList v1.List
+	if err := yaml.Unmarshal(b, &sourceList); err != nil {
+		return nil, err
+	}
+
+	items := []runtime.RawExtension{}
+
+	for _, item := range sourceList.Items {
+		result, err := injectResource(item.Raw, version)
+		if err != nil {
+			return nil, err
 		}
 
-		// If we don't inject anything into the pod template then output the
-		// original serialization of the original object. Otherwise, output the
-		// serialization of the modified object.
-		output := bytes
-		if podTemplateSpec != nil && injectPodTemplateSpec(podTemplateSpec, DNSNameOverride, version, k8sLabels) {
-			output, err = yaml.Marshal(obj)
-			if err != nil {
-				return err
-			}
+		injected, err := yaml.YAMLToJSON(result)
+		if err != nil {
+			return nil, err
 		}
 
-		out.Write(output)
-		out.Write([]byte("---\n"))
+		items = append(items, runtime.RawExtension{Raw: injected})
 	}
-	return nil
+
+	sourceList.Items = items
+	return yaml.Marshal(sourceList)
+}
+
+func injectResource(bytes []byte, version string) ([]byte, error) {
```

Comment:
can you add a comment indicating that if there's nothing to inject, this function will return the same pointer value as the input?

another option: always return an object distinct from the input.